### ISSUE TITLE
Assignment of swizzled values should be permitted

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#737]
It's really common to see things like like `v.xz = float2(0, 1);` in shaders. We should try to find a way for this to work.

**Comment by dneto0** [comment]
To be clear, the single-letter form is intended to be supported already. The multi-letter swizzle as l-value is explicitly called out in a note as not supported.

I think WGSL should have the equivalent of OpenCL C's ["shuffle2"](https://www.khronos.org/registry/OpenCL/sdk/2.1/docs/man/xhtml/shuffle.html), which has a close analog in LLVM IR in as "shufflevector" and in SPIR-V as OpVectorShuffle.
The operate on vectors as values, not as a vector-valued variable (storage).

Assuming v is a 4-element vector, Glslang implements   `v.xz = foo` in SPIR-V   as the following:

     %initial = load from v
     %newvalue = OpVectorShuffle %v_type %initial %foo 4 1 5 3
     OpStore %v %newvalue

**Comment by dneto0** [comment]
I separated out the shuffle primitive to #748

**Comment by johnkslang** [comment]
Agreed with separating these, and for terminology, it will help to not include the single-letter forms as "swizzles".  The single letter forms map directly to HLL `[]` concepts and to SPIR-V **Extract/Insert** instructions.

It is the multi-letter versions that are trickier to sort out and having unique ties to how graphics work.  It will help to have those map 1:1 with swizzle terminology.

**Comment by kvark** [comment]
I think we can live without multi-element left-hand-side swizzles, at least postponing this until after MVP.

**Comment by grorg** [comment]
Discussed at the [2020-05-26 meeting](https://docs.google.com/document/d/1KuWVqle7kNaQgjcPxk_iqBZYCbgP9yIg_Ek7t7D8k3w/).

Resolution:
- `rgba` and `xywz`
- they mean the same thing
- no to `stpq`
- you can’t mix the swizzles e.g. `rybz` is a compilation error
- vectors only

**Comment by litherum** [comment]
In the HLSL corpus, 0.1% of lines assign to multi-element swizzles, so we can defer this topic until later.

**Comment by cwfitzgerald** [comment]
To throw another data point in the fire, this is one of the most common questions I get about wgsl  as for "why my code doesn't compile". The answer is basically "you can't do that because we said so" which isn't very satisfying for anyone. I'm fine if y'all want to push this post mvp, but let's not forget about it.

**Comment by munrocket** [comment]
Maybe we can modify this statement
```ts
p.xz = f(w);
```
into this
```ts
let temp = f(w); p.x = temp.x; p.z = temp.y;
```
and compile with the same codegen.

**Comment by johnstiles-google** [comment]
> Maybe we can modify this statement
> 
> ```ts
> p.xz = f(w);
> ```
> 
> into this
> 
> ```ts
> let temp = f(w); p.x = temp.x; p.z = temp.y;
> ```
> 
> and compile with the same codegen.

That's unlikely to match in terms of codegen. Swizzle assignment in SPIR-V works by jamming together two inputs and applying a swizzle to the concatenation of the two values: 

  %50 = OpVectorShuffle %v4float %49 %48 0 4 5 6

Which means %50 = vec4(%49.x, %48.xyz)

**Comment by johnstiles-google** [comment]
On a semi-related note, it's possible to write code in other shading languages which combines swizzles and indexes on an lvalue, such as:

`myVec4.wzyx[index] = 0;`

This would clear `myVec4.w` if index is 0, or `myVec4.z` if index is 1, etc.

However, in practice this can be dicey—in Metal it doesn't always work properly. I've seen cases where it can corrupt neighboring stack entries. Some GLSL implementations also struggle to get it right. YMMV. 

We should probably make sure to spec the WGSL version of this feature clearly, so we don't end up dragging in adjacent buggy features that we don't need.

**Comment by magcius** [comment]
> That's unlikely to match in terms of codegen.

What is unlikely to match here? The semantics are the same, no?

> On a semi-related note, it's possible to write code in other shading languages which combines swizzles and indexes on an lvalue

I was not aware this was possible, and it is not important for my use case.

**Comment by johnstiles-google** [comment]
> What is unlikely to match here? The semantics are the same, no?

Yes, semantically it's the same meaning, but it's expressed in a less efficient way. I don't know if the optimizer would be smart enough to optimize the one-field-at-a-time case back into the all-fields-at-once case. If performance is not a concern, then it's fine, but performance is always a concern :) 

> I was not aware this was possible, and it is not important for my use case.

100% agree—I'm just trying to call out that this is a thing, and WGSL probably doesn't want to allow it, but it's easy to accidentally write the spec in such a way that we do allow it and then we are stuck with it.

**Comment by magcius** [comment]
https://godbolt.org/z/Pn71qfd3c

Seems identical codegen to me. This is admittedly HLSL/DXC, not SPIR-V, but I would expect the same thing out of SPIR-V as well. SPIR-V Vectors are an illusion anyway :)

**Comment by myth0genesis** [comment]
I'd like to see assignment of multi-valued swizzles, as well. I don't think looking at the percentage of lines in shaders using a given functionality is a good metric to determine if something warrants attention. I probably only use it in about 0.1% of my lines, but I use it in more than 95% of my shaders, as I use 2D rotation matrices in almost all of them, and the most concise way to represent them is with a two-value swizzle-assignment:

```
fn rot2D(a: f32) -> mat2x2f {
  var s: f32 = sin(a);
  var c: f32 = cos(a);
  return mat2x2f(c, -s,
                 s,  c);
}

@fragment
fn fragmentMain(@builtin(position) pos: vec4f) -> @location(0) vec4f {
...
  var ro: vec3f = vec3f(0.0, 3.0, -7.0);

  // This would be nice:
  ro.xz *= rot2D(0.5);
...
}
```
The relevant line:
```
  ro.xz *= rot2D(0.5);
```
is a lot easier to understand than:
```
  ro = vec3f(ro.xz * rot2D(0.5), ro.y).xzy;
```
if you wanted to achieve the same thing in a single line.

**Comment by dneto0** [comment]
@tomforsyth1000 on mastodon:

https://mastodon.gamedev.place/@TomF/111718774130914484 

> No - I completely agree with their decision. Swizzled assignment is the source of a gigantic number of bugs. Depending on your shader language (and version) of choice:

> p3.yz = r.xy;

> ...may do an exciting variety of things. Total bug farm.

**Comment by myth0genesis** [comment]
How so? I've used that feature in nearly every shader I've written and not once have I ever experienced a bug because of it. If you know a particular implementation that is causing an undesired behavior, like padding a 4-tuple by repeating the last letter as you mentioned in your Mastodon post, then don't implement it like that?

**Comment by kainino0x** [comment]
I basically agree, if there are driver bugs then we should work around the driver bugs - that's what we do :)

**Comment by munrocket** [comment]
People asking swizzling for milestone-1 in wgsl. Lot of thumbs up here.

**Comment by Waffle1434** [comment]
This is definitely one of the more painful parts of WGSL right now. Which is surprising, since it got a lot of the HLSL syntax convenience features but missed this one.

**Comment by dneto0** [comment]
Fixed by #5268

See also #732 for swizzle syntax  (in r-values) that allows 0 and 1 in components.
